### PR TITLE
Adding Variant Sender

### PR DESCRIPTION
**Pull request by MarkCMann**
Returning multiple senders from within a `let_value*` function requires either type erasing the results or inventing new types to represent the multiple senders. This is cumbersome since type erasure isn't always needed. 

This PR introduces the VariantSender type that wraps an underlying sender and provides an easy mechanism to create new Sender types that collapse multiple other Senders.

### CODE REVIEW

File: include/unifex/variant_sender.hpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    noexcept(std::is_nothrow_constructible_v<std::variant<Senders...>, decltype(concrete_sender)>)
+    : sender_variant(std::forward<ConcreteSender>(concrete_sender)) {}
+
+  template(typename This, typename Receiver)
+    (requires same_as<remove_cvref_t<This>, type> AND std::conjunction_v<std::bool_constant<sender_to<member_t<This, Senders>, Receiver>>...>)
+  friend auto tag_invoke(tag_t<connect>, This&& that, Receiver&& r)
+    noexcept(std::conjunction_v<unifex::is_nothrow_connectable<Senders, Receiver>...>)
+  {
+    return operation<connect_result_t<Senders, Receiver>...>{
+        std::visit([r = static_cast<Receiver&&>(r)](auto&& sender) mutable noexcept(unifex::is_nothrow_connectable_v<decltype(sender), Receiver>) {
```
Comment (ispeters) [inline include/unifex/variant_sender.hpp:108]: This lambda is executed synchronously so I think you can capture `r` by reference and forward directly into the nested `connect` instead of forwarding and then moving.
```suggestion
        std::visit([&r](auto&& sender) noexcept(unifex::is_nothrow_connectable_v<decltype(sender), Receiver>) {
```

File: include/unifex/variant_sender.hpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            return std::variant<connect_result_t<Senders, Receiver>...>{unifex::connect(std::move(sender), static_cast<Receiver&&>(r))};
+        }, std::move(static_cast<This&&>(that).sender_variant))
```
Comment (ispeters) [inline include/unifex/variant_sender.hpp:110]: I think you're moving the sender when you should be forwarding it here.
```suggestion
            return std::variant<connect_result_t<Senders, Receiver>...>{unifex::connect(static_cast<decltype(sender)>(sender), static_cast<Receiver&&>(r))};
        }, static_cast<This&&>(that).sender_variant)
```

File: include/unifex/variant_sender.hpp [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+  template<typename ConcreteSender>
+  type(ConcreteSender&& concrete_sender)
+    noexcept(std::is_nothrow_constructible_v<std::variant<Senders...>, decltype(concrete_sender)>)
+    : sender_variant(std::forward<ConcreteSender>(concrete_sender)) {}
+
+  template(typename This, typename Receiver)
+    (requires same_as<remove_cvref_t<This>, type> AND std::conjunction_v<std::bool_constant<sender_to<member_t<This, Senders>, Receiver>>...>)
+  friend auto tag_invoke(tag_t<connect>, This&& that, Receiver&& r)
+    noexcept(std::conjunction_v<unifex::is_nothrow_connectable<Senders, Receiver>...>)
```
Comment (ispeters) [inline include/unifex/variant_sender.hpp:105]: I think this `noexcept` clause is imprecise.  I think the member variant ought to be forwarded into `visit`, which means the `connect` expression may be given either an lvalue or an rvalue for any of the constituent senders.

I wonder what the value categories end up being in the cross product of (`This` is an value, `This` is an rvalue) x (one of `Senders` is a plain type, one of `Senders` is an lvalue reference, one of `Senders` is an rvalue reference).

**Review by ispeters — CHANGES_REQUESTED** [review]
Looks good, but I have some questions about `noexcept` and value categories.

Can you add some test cases to cover the weird value category corner cases, and at least some minimal documentation to the API reference, please?

File: include/unifex/variant_sender.hpp (hunk 3 above)
Comment (MarkCMann) [reply include/unifex/variant_sender.hpp:105]: Seems that [`std::variant` isn't permitted to hold references](https://en.cppreference.com/w/cpp/utility/variant):
> A variant is not permitted to hold references, arrays, or the type void. Empty variants are also ill-formed (std::variant<[std::monostate](http://en.cppreference.com/w/cpp/utility/variant/monostate)> can be used instead).

So I'll add some tests for when `This` is an lvalue/rvalue.

File: test/variant_sender_test.cpp [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+    void set_error(std::exception_ptr) noexcept {}
+};
+
+template<bool lvalueConnectNoexcept = true, bool rvalueConnectNoexcept = true>
+struct TestSender {
+  template <
+    template <typename...> class Variant,
+    template <typename...> class Tuple>
+  using value_types = Variant<std::string>;
```
Comment (ispeters) [inline test/variant_sender_test.cpp:55]: I think you need to use `Tuple` somewhere to make this compatible with all the type machinery, no?
```suggestion
  using value_types = Variant<Tuple<std::string>>;
```

File: test/variant_sender_test.cpp [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+template<bool lvalueConnectNoexcept, bool rvalueConnectNoexcept>
+using test_sender_t = variant_sender<TestSender<lvalueConnectNoexcept, rvalueConnectNoexcept>>;
+
+template<typename T, bool lvalue>
+using conditionally_lvalue_t = std::conditional_t<lvalue, std::add_lvalue_reference_t<T>, T>;
+
+template<bool lvalueConnectNoexcept, bool rvalueConnectNoexcept, bool isLvalueReference = true>
+using is_noexcept = unifex::is_nothrow_connectable<conditionally_lvalue_t<test_sender_t<lvalueConnectNoexcept, rvalueConnectNoexcept>, isLvalueReference>, IntAndStringReceiver>;
```
Comment (ispeters) [inline test/variant_sender_test.cpp:176]: These should be in an anonymous namespace, I think.
```suggestion
namespace {

template<bool lvalueConnectNoexcept, bool rvalueConnectNoexcept>
using test_sender_t = variant_sender<TestSender<lvalueConnectNoexcept, rvalueConnectNoexcept>>;

template<typename T, bool lvalue>
using conditionally_lvalue_t = std::conditional_t<lvalue, std::add_lvalue_reference_t<T>, T>;

template<bool lvalueConnectNoexcept, bool rvalueConnectNoexcept, bool isLvalueReference = true>
using is_noexcept = unifex::is_nothrow_connectable<conditionally_lvalue_t<test_sender_t<lvalueConnectNoexcept, rvalueConnectNoexcept>, isLvalueReference>, IntAndStringReceiver>;

} // namespace
```

File: test/variant_sender_test.cpp [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+  static constexpr bool sends_done = true;
+
+  struct op {};
+
+  template<typename Receiver>
+  op connect(Receiver&& r) & noexcept(lvalueConnectNoexcept);
+
+  template<typename Receiver>
+  op connect(Receiver&& r) && noexcept(rvalueConnectNoexcept);
+};
```
Comment (ispeters) [inline test/variant_sender_test.cpp:69]: I think these should be in an anonymous namespace.
```suggestion
namespace {

struct IntAndStringReceiver {
    void set_value(int) noexcept {

    }

    void set_value(std::string) noexcept {

    }

    void set_done() noexcept {}

    void set_error(std::exception_ptr) noexcept {}
};

template<bool lvalueConnectNoexcept = true, bool rvalueConnectNoexcept = true>
struct TestSender {
  template <
    template <typename...> class Variant,
    template <typename...> class Tuple>
  using value_types = Variant<std::string>;

  template <template <typename...> class Variant>
  using error_types = Variant<>;

  static constexpr bool sends_done = true;

  struct op {};

  template<typename Receiver>
  op connect(Receiver&& r) & noexcept(lvalueConnectNoexcept);

  template<typename Receiver>
  op connect(Receiver&& r) && noexcept(rvalueConnectNoexcept);
};

} // namespace
```

**Review by ispeters — APPROVED** [review]
Some nits; I think you can squash and merge after you address them.
